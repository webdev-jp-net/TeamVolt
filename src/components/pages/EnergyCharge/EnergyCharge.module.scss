@use "style/_variable" as *;

.article {
  display: grid;
  grid-template-rows: auto 1fr auto;
  height: 100%;
  padding: rem(48) 0 0;

  * {
    user-select: none;
  }
}

// ヘッダー
.header {
  display: flex;
  flex-direction: column;
  gap: rem(24);
  margin: rem(24) rem(24) auto;
}

// テキスト
.paragraph {
  margin: rem(24) 0;
}

.body {
  padding: 0 rem(24);
  margin: rem(48) 0 0;
}

// 準備
.ready {
  display: flex;
  flex-direction: column;
  gap: rem(24);
  margin: rem(24) 0;
}

.console {
  display: flex;
  gap: rem(12);
  align-items: center;
  // margin: rem(24) 0;
}

.result {
  margin-top: rem(-8);
  font-size: rem(18);
}

// stock
.visualizer {
  position: absolute;
  top: rem(120);
  right: rem(24);
  left: rem(24);
  display: grid;
  align-items: end;
}

.stock {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  gap: rem(24) 0;
  align-items: end;
  height: 1;
  margin: rem(36) 0;
  font-size: rem(72);
}

.stockItem {
  width: rem(72);
  line-height: 1;
}

.gauge {
  margin: 0 rem(12);

  &.--invisible {
    opacity: 0;
  }
}

// タップエリア
.tap {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: rem(180) 0 0;
  color: $c_dark;
  background: transparent;
  border: none;
  -webkit-tap-highlight-color: transparent;
}

// 結果確認
.hasChallenged {
  display: flex;
  flex-direction: column;
  gap: rem(24);
  padding: 0 rem(24);
  margin: rem(48) 0 0;
}

// フッター
.footer {
  display: flex;
  flex-direction: column;
  gap: rem(24);
  margin: auto rem(24) rem(24);
}
