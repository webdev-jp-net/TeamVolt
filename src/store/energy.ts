import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import { doc, getDoc, updateDoc, arrayUnion, runTransaction } from 'firebase/firestore';
import { db } from 'firebaseDB';

type ChargeUnitsData = { member: string; count: number };

type State = {
  genEnergy: number;
  chargeUnits: ChargeUnitsData[];
};
const initialState: State = {
  genEnergy: 0,
  chargeUnits: [],
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview

export const energyGetApi = createApi({
  reducerPath: 'energyGetApi',
  baseQuery: async id => {
    if (id) {
      const docRef = doc(db, 'team', id);
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: docSnap.data().chargeUnits };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // 獲得バッテリー取得
    getChargeUnits: builder.query<ChargeUnitsData[], string>({
      query: id => id,
    }),
  }),
});
export const { useGetChargeUnitsQuery } = energyGetApi;

export const energyPostApi = createApi({
  reducerPath: 'energyPostApi',
  baseQuery: async ({ id, value }: { id?: string; value: ChargeUnitsData }) => {
    if (id && value) {
      const docRef = doc(db, 'team', id);
      try {
        await runTransaction(db, async transaction => {
          const sfDoc = await transaction.get(docRef);
          if (!sfDoc.exists()) {
            // eslint-disable-next-line no-throw-literal
            throw 'Document does not exist!';
          }
          const oldList = sfDoc.data().chargeUnits || [];
          const newList = [
            ...oldList.filter((item: ChargeUnitsData) => item.member !== value.member),
            value,
          ];
          transaction.update(docRef, { chargeUnits: newList });
        });
      } catch (e) {
        console.log('Transaction failure:', e);
      }
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: docSnap.data().chargeUnits };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // 獲得バッテリー追加
    addChargeUnits: builder.mutation<ChargeUnitsData[], { id: string; value: ChargeUnitsData }>({
      query: ({ id, value }) => ({
        operationType: 'remove_challenger',
        id,
        value,
      }),
    }),
  }),
});

export const { useAddChargeUnitsMutation } = energyPostApi;

const energy = createSlice({
  name: 'energy',

  initialState,

  reducers: {
    // 獲得エネルギーを更新
    updateGenEnergy: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        genEnergy: action.payload,
      };
    },
    // 獲得エネルギーをリセット
    resetEnergy: () => {
      return {
        genEnergy: 0,
        chargeUnits: [],
      };
    },
  },

  extraReducers: builder => {
    // 成功: 獲得バッテリー取得
    builder.addMatcher(
      energyGetApi.endpoints.getChargeUnits.matchFulfilled,
      (state, action: PayloadAction<ChargeUnitsData[]>) => {
        const result = { chargeUnits: action.payload };
        return {
          ...state,
          ...result,
        };
      }
    );
    // 成功: 獲得バッテリー追加
    builder.addMatcher(
      energyPostApi.endpoints.addChargeUnits.matchFulfilled,
      (state, action: PayloadAction<ChargeUnitsData[]>) => {
        const result = { chargeUnits: action.payload };
        return {
          ...state,
          ...result,
        };
      }
    );
  },
});

// Action Creator
export const { updateGenEnergy, resetEnergy } = energy.actions;

// Reducer
export default energy.reducer;
