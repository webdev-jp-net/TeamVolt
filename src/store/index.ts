import { combineReducers, configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';

import user, { statusGetApi, statusUpdateApi } from './user';

const reducer = combineReducers({
  user,
  [statusGetApi.reducerPath]: statusGetApi.reducer,
  [statusUpdateApi.reducerPath]: statusUpdateApi.reducer,
});

const middleware = getDefaultMiddleware({ serializableCheck: false }).concat(
  statusGetApi.middleware,
  statusUpdateApi.middleware
);

export const store = configureStore({ reducer, middleware });

export type RootState = ReturnType<typeof reducer>;
