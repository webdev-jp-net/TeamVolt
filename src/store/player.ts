import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import {
  doc,
  getDoc,
  getDocs,
  updateDoc,
  arrayRemove,
  arrayUnion,
  addDoc,
  deleteField,
  collection,
  runTransaction,
} from 'firebase/firestore';
import { db } from 'firebaseDB';

import type { PlayerArticleData } from 'types/player';
import type { TeamArticleData, ChargeUnitsData, MemberData } from 'types/team';

type State = {
  localId: string;
  player?: string;
  playerList: PlayerArticleData[];
  selectedTeam?: string;
  myTeam?: TeamArticleData;
  genEnergy: number;
};

const initialState: State = {
  localId: '',
  playerList: [],
  genEnergy: 0,
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// 読込用
export const playerGetApi = createApi({
  reducerPath: 'playerGetApi',
  baseQuery: async () => {
    const playerRef = collection(db, 'player');
    const playerSnapshot = await getDocs(playerRef);
    const result = playerSnapshot.docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
    return { data: result };
  },
  endpoints: builder => ({
    // プレイヤー情報取得
    getPlayer: builder.query<PlayerArticleData[], void>({
      query: () => ({}),
    }),
  }),
});
export const { useGetPlayerQuery } = playerGetApi;

// チーム情報取得
export const teamArticleGetApi = createApi({
  reducerPath: 'teamArticleGetApi',
  baseQuery: async ({ id }) => {
    const docRef = doc(db, 'team', id);
    const docSnap = await getDoc(docRef);
    if (docSnap.exists()) {
      return { data: { id, ...docSnap.data() } };
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // チーム情報取得
    getTeamArticle: builder.query<TeamArticleData, string>({
      query: id => ({ id }),
    }),
  }),
});
export const { useGetTeamArticleQuery } = teamArticleGetApi;

// 書込用
export const playerAddApi = createApi({
  reducerPath: 'playerAddApi',
  baseQuery: async ({ value }) => {
    const playerRef = collection(db, 'player');
    const docRef = await addDoc(playerRef, {
      localId: value,
    });
    return { data: { id: docRef.id, localId: value } };
  },
  endpoints: builder => ({
    // プレイヤー情報追加
    addPlayer: builder.mutation<PlayerArticleData, string>({
      query: value => ({ value }),
    }),
  }),
});
export const { useAddPlayerMutation } = playerAddApi;

// チーム情報の更新

// チームメンバー編集
type teamPostApiProps = {
  operationType: string;
  id?: string;
  value?: string | number | ChargeUnitsData | MemberData;
};
export const teamPostApi = createApi({
  reducerPath: 'teamPostApi',
  baseQuery: async ({ operationType, id, value }: teamPostApiProps) => {
    if (operationType === 'add_member' && id) {
      // メンバーを追加
      const docRef = doc(db, 'team', id);
      await updateDoc(docRef, {
        member: arrayUnion(value),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'remove_member' && id) {
      const docRef = doc(db, 'team', id);
      // メンバーを削除
      await updateDoc(docRef, {
        member: arrayRemove(value),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'add_challenger' && id) {
      const docRef = doc(db, 'team', id);
      // 代表者を登録
      await updateDoc(docRef, {
        challenger: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'remove_challenger' && id) {
      const docRef = doc(db, 'team', id);
      // 代表者を削除
      await updateDoc(docRef, {
        challenger: deleteField(),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'add_charge_units' && id && value) {
      // 獲得バッテリーを登録
      const docRef = doc(db, 'team', id);
      try {
        await runTransaction(db, async transaction => {
          const sfDoc = await transaction.get(docRef);
          if (!sfDoc.exists()) {
            // eslint-disable-next-line no-throw-literal
            throw 'Document does not exist!';
          }
          const oldList = sfDoc.data().chargeUnits || [];
          const newList = [
            ...oldList.filter(
              (item: ChargeUnitsData) =>
                item.member !== (value as unknown as ChargeUnitsData).member
            ),
            value,
          ];
          transaction.update(docRef, { chargeUnits: newList });
        });
      } catch (e) {
        console.log('Transaction failure:', e);
      }
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'update_used_units' && id) {
      const docRef = doc(db, 'team', id);
      // 救出ミッション試行回数を登録
      await updateDoc(docRef, {
        usedUnits: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'update_current_positions' && id) {
      const docRef = doc(db, 'team', id);
      // 救出ミッション進捗を登録
      await updateDoc(docRef, {
        currentPosition: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'close_mission' && id) {
      const docRef = doc(db, 'team', id);
      // メンバー、役割、プレイ結果を削除
      await updateDoc(docRef, {
        member: deleteField(),
        challenger: deleteField(),
        chargeUnits: deleteField(),
        currentPosition: deleteField(),
        usedUnits: deleteField(),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // メンバー追加
    addMember: builder.mutation<TeamArticleData, { id: string; value: MemberData }>({
      query: ({ id, value }) => ({
        operationType: 'add_member',
        id,
        value,
      }),
    }),
    // メンバー削除
    removeMember: builder.mutation<TeamArticleData, { id: string; value: MemberData }>({
      query: ({ id, value }) => ({
        operationType: 'remove_member',
        id,
        value,
      }),
    }),
    // 抽選結果追加
    addChallenger: builder.mutation<TeamArticleData, { id: string; value: string }>({
      query: ({ id, value }) => ({
        operationType: 'add_challenger',
        id,
        value,
      }),
    }),
    // 抽選結果追加
    removeChallenger: builder.mutation<TeamArticleData, { id: string; value: string }>({
      query: ({ id, value }) => ({
        operationType: 'remove_challenger',
        id,
        value,
      }),
    }),
    // 獲得バッテリー追加
    addChargeUnits: builder.mutation<TeamArticleData, { id: string; value: ChargeUnitsData }>({
      query: ({ id, value }) => ({
        operationType: 'add_charge_units',
        id,
        value,
      }),
    }),
    // 救出ミッション試行数を更新
    updateUsedUnits: builder.mutation<TeamArticleData, { id: string; value: number }>({
      query: ({ id, value }) => ({
        operationType: 'update_used_units',
        id,
        value,
      }),
    }),
    // 救出ミッション進捗を更新
    updateCurrentPosition: builder.mutation<TeamArticleData, { id: string; value: number }>({
      query: ({ id, value }) => ({
        operationType: 'update_current_positions',
        id,
        value,
      }),
    }),
    // ミッション情報をクリア
    closeMission: builder.mutation<TeamArticleData, string>({
      query: id => ({
        operationType: 'close_mission',
        id,
      }),
    }),
  }),
});
export const {
  useAddMemberMutation,
  useRemoveMemberMutation,
  useAddChallengerMutation,
  useRemoveChallengerMutation,
  useAddChargeUnitsMutation,
  useUpdateUsedUnitsMutation,
  useUpdateCurrentPositionMutation,
  useCloseMissionMutation,
} = teamPostApi;

const player = createSlice({
  name: 'player',

  initialState,

  reducers: {
    // ローカルID更新
    updateLocalId: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        localId: action.payload,
      };
    },
    // 所属チーム更新
    updateTeam: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        selectedTeam: action.payload,
      };
    },
    // 所属チーム脱退
    escapeTeam: state => {
      return {
        ...state,
        selectedTeam: undefined,
      };
    },
    // 獲得エネルギーを更新
    updateGenEnergy: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        genEnergy: action.payload,
      };
    },
    // 獲得エネルギーをリセット
    resetEnergy: store => {
      return {
        ...store,
        genEnergy: 0,
      };
    },
  },
  extraReducers: builder => {
    // 成功: プレイヤー情報取得
    builder.addMatcher(
      playerGetApi.endpoints.getPlayer.matchFulfilled,
      (state, action: PayloadAction<PlayerArticleData[]>) => {
        state.playerList = action.payload;
      }
    );
    // 成功: プレイヤー追加
    builder.addMatcher(
      playerAddApi.endpoints.addPlayer.matchFulfilled,
      (state, action: PayloadAction<PlayerArticleData>) => {
        const result = { playerList: [...state.playerList, action.payload] };
        return {
          ...state,
          ...result,
        };
      }
    );
    // 成功: チーム情報取得
    builder.addMatcher(
      teamArticleGetApi.endpoints.getTeamArticle.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: メンバー追加
    builder.addMatcher(
      teamPostApi.endpoints.addMember.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: メンバー削除
    builder.addMatcher(
      teamPostApi.endpoints.removeMember.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 代表者追加
    builder.addMatcher(
      teamPostApi.endpoints.addChallenger.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 代表者削除
    builder.addMatcher(
      teamPostApi.endpoints.removeChallenger.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 獲得バッテリー追加
    builder.addMatcher(
      teamPostApi.endpoints.addChargeUnits.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 救出ミッション試行数更新
    builder.addMatcher(
      teamPostApi.endpoints.updateUsedUnits.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 救出ミッション進捗更新
    builder.addMatcher(
      teamPostApi.endpoints.updateCurrentPosition.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: ミッション情報クリア成功
    builder.addMatcher(
      teamPostApi.endpoints.closeMission.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
  },
});

// Action Creator
export const { updateLocalId, updateTeam, escapeTeam, updateGenEnergy, resetEnergy } =
  player.actions;

// Reducer
export default player.reducer;
