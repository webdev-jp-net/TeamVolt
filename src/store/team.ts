import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import { getDocs, addDoc, collection } from 'firebase/firestore';
import { db } from 'firebaseDB';
import { TeamArticleData } from 'types/team';

type State = {
  teamList: TeamArticleData[];
};

const initialState: State = {
  teamList: [],
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// read
export const teamGetApi = createApi({
  reducerPath: 'teamGetApi',
  baseQuery: async ({ path }) => {
    const teamRef = collection(db, path);
    const teamSnapshot = await getDocs(teamRef);
    const result = teamSnapshot.docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
    return { data: result };
  },
  endpoints: builder => ({
    // チーム情報取得
    getTeam: builder.query<TeamArticleData[], void>({
      query: () => ({ path: 'team' }),
    }),
  }),
});
export const { useGetTeamQuery } = teamGetApi;

// write
export const teamAddApi = createApi({
  reducerPath: 'teamAddApi',
  baseQuery: async ({ value }) => {
    const teamRef = collection(db, 'team');
    const docRef = await addDoc(teamRef, {
      name: value,
      member: [],
    });
    return { data: { id: docRef.id, name: value, member: [] } };
  },
  endpoints: builder => ({
    // チーム情報追加
    addTeam: builder.mutation<TeamArticleData, string>({
      query: value => ({ value }),
    }),
  }),
});
export const { useAddTeamMutation } = teamAddApi;

const team = createSlice({
  name: 'team',

  initialState,

  reducers: {
    updateTeamList: (state, action: PayloadAction<TeamArticleData[]>) => {
      return {
        ...state,
        teamList: action.payload,
      };
    },
  },

  extraReducers: builder => {
    // 成功: チーム情報取得
    builder.addMatcher(
      teamGetApi.endpoints.getTeam.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData[]>) => {
        state.teamList = action.payload;
      }
    );
    // 成功: チーム追加
    builder.addMatcher(
      teamAddApi.endpoints.addTeam.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        const result = { teamList: [...state.teamList, action.payload] };
        return {
          ...state,
          ...result,
        };
      }
    );
  },
});

// Action Creator
export const { updateTeamList } = team.actions;

// Reducer
export default team.reducer;
