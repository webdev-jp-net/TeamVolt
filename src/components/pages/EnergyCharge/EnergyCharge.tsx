import { FC, useState, useEffect, useMemo, useCallback } from 'react';

import { useSelector, useDispatch } from 'react-redux';

import { useNavigate } from 'react-router-dom';

import { Button } from 'components/parts/Button';
import { RootState } from 'store';
import { updateGenEnergy, resetEnergy } from 'store/player';
import { useGetTeamArticleQuery, useAddChargeUnitsMutation } from 'store/team';

import styles from './EnergyCharge.module.scss';

import { GaugeUi } from './components/GaugeUi';
import { TimeLeftUi } from './components/TimeLeftUi';

export const EnergyCharge: FC = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { localId, genEnergy } = useSelector((state: RootState) => state.player);
  const { teamList, selectedTeam } = useSelector((state: RootState) => state.team);

  // 所属チームの情報
  const myTeam = useMemo(() => {
    return teamList.find(team => team.id === selectedTeam);
  }, [teamList, selectedTeam]);

  // 役職
  const job = useMemo(() => {
    return myTeam?.challenger === localId ? 'Rescuer' : 'Charger';
  }, [myTeam, localId]);

  // チャレンジ済みフラグ
  const hasChallenged = useMemo(() => {
    return myTeam?.chargeUnits
      ? myTeam?.chargeUnits.some(chargeUnit => chargeUnit.member === localId)
      : false;
  }, [myTeam, localId]);

  // 準備完了フラグ
  const [isReady, setIsReady] = useState<boolean>(false);

  // 残り時間
  const [currentTime, setCurrentTime] = useState<number>(0);
  const limit = 30;
  const chargeThreshold = 50;

  // カウントダウンの処理
  useEffect(() => {
    let timerID: NodeJS.Timeout | undefined;
    if (isReady && currentTime < limit) {
      timerID = setInterval(() => {
        setCurrentTime(currentTime => currentTime + 1);
      }, 1000);
    } else {
      clearInterval(timerID!);
    }
    return () => {
      clearInterval(timerID!);
    };
  }, [isReady, currentTime]);

  // スコアを加算
  const handleTap = () => {
    dispatch(updateGenEnergy(genEnergy + 1));
  };

  // 獲得バッテリーを送信
  const [sendAddChargeUnits] = useAddChargeUnitsMutation();
  const handleSubmit = useCallback(() => {
    const count = Math.floor(genEnergy / chargeThreshold);
    sendAddChargeUnits({ id: selectedTeam || '', value: { member: localId, count } });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [genEnergy, selectedTeam, localId]);

  // チーム情報の取得
  const {
    isLoading: getDrawResultLoading,
    refetch: getDrawResultRefetch,
    isFetching: getDrawResultFetching,
  } = useGetTeamArticleQuery(selectedTeam || '', { skip: !selectedTeam });

  const handleTeamStockRequest = useCallback(() => {
    getDrawResultRefetch();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // チームで獲得しているバッテリー数
  const totalChargeUnits = useMemo(() => {
    return myTeam?.chargeUnits
      ? myTeam?.chargeUnits.reduce((acc, chargeUnit) => {
          return acc + chargeUnit.count;
        }, 0)
      : 0;
  }, [myTeam]);

  // 離脱のタイミングでリセット
  useEffect(() => {
    document.body.classList.add('isFixedScroll');
    return () => {
      document.body.classList.remove('isFixedScroll');
      dispatch(resetEnergy());
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <article className={styles.article}>
      <header className={styles.header}>
        <h1>Energy Charge</h1>
        <p>
          You're on <strong>{job}</strong> duty!
        </p>
      </header>
      {job === 'Rescuer' || hasChallenged ? (
        <>
          <div className={styles.hasChallenged}>
            <Button
              handleClick={handleTeamStockRequest}
              disabled={!selectedTeam || getDrawResultLoading || getDrawResultFetching}
            >
              Check Team Charge
            </Button>
            <div className={styles.stock}>
              {Array(totalChargeUnits)
                .fill(0)
                .map((_, index) => (
                  <span key={index} className={styles.stockItem}>
                    🔋
                  </span>
                ))}
            </div>
          </div>
          <footer className={styles.footer}>
            <Button
              handleClick={() => {
                navigate('/target-lead');
              }}
              disabled={totalChargeUnits > 0}
            >
              Start Rescue
            </Button>
          </footer>
        </>
      ) : (
        <>
          <div className={styles.body}>
            {!isReady ? (
              <div className={styles.ready}>
                <p>Tap the screen many times to charge the battery.</p>
                <Button
                  handleClick={() => {
                    setIsReady(true);
                  }}
                >
                  I'm Ready
                </Button>
              </div>
            ) : currentTime < limit ? (
              <>
                <div className={styles.console}>
                  <TimeLeftUi currentTime={currentTime} limit={limit} />
                  <p className={styles.genEnergy}>⚡️ {genEnergy}</p>
                </div>
                <button type="button" onClick={handleTap} className={styles.tap}>
                  Tap Screen!!
                </button>
              </>
            ) : (
              <>
                <h2 className={styles.result}>Batteries Earned</h2>
                <p>Great job! Send your results to the Rescuers.</p>
              </>
            )}
            <div className={styles.stock}>
              {Array(Math.floor(genEnergy / chargeThreshold))
                .fill(0)
                .map((_, index) => (
                  <span key={index} className={styles.stockItem}>
                    🔋
                  </span>
                ))}
              {genEnergy > 0 && (
                <GaugeUi
                  addClass={[styles.gauge, currentTime >= limit ? styles['--invisible'] : '']}
                  currentValue={Math.floor(((genEnergy % chargeThreshold) / chargeThreshold) * 100)}
                />
              )}
            </div>
          </div>
          <footer className={styles.footer}>
            {currentTime >= limit && (
              <>
                <Button handleClick={handleSubmit}>submit</Button>
              </>
            )}
          </footer>
        </>
      )}
    </article>
  );
};
