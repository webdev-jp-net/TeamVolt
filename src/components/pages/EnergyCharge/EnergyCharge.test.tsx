import { Router } from 'react-router-dom';

import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { store } from 'store';

import { EnergyCharge } from './EnergyCharge';
const history = createMemoryHistory();

describe('EnergyCharge', () => {
  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  it('EnergyChargeのスナップショット', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <EnergyCharge />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
