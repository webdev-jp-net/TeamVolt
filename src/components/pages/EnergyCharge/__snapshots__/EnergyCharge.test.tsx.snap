// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnergyCharge EnergyChargeのスナップショット 1`] = `
<DocumentFragment>
  <article
    class="article"
  >
    <header
      class="header"
    >
      <h1
        class="title"
      >
        あなたは 
        <strong
          class="job"
        >
          <svg
            class="jobIcon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M14.5 11l-3 6v-4h-2l3-6v4h2zM7 1h10c1.1 0 2 .9 2 2v18c0 1.1-.9 2-2 2H7c-1.1 0-2-.9-2-2V3c0-1.1.9-2 2-2zm0 5v12h10V6H7z"
            />
          </svg>
          充電係
        </strong>
         です！
      </h1>
    </header>
    <div
      class="body"
    >
      <div
        class="challengeConsole"
      >
        <div
          class="timeLeftUi "
        >
          <svg
            class="gauge"
            version="1.1"
            viewBox="0 0 66 66"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M33,0c-0.6,0-1,0.4-1,1v4c0,0.6,0.4,1,1,1s1-0.4,1-1V1C34,0.4,33.6,0,33,0z M12.5,14.1c0.4,0.4,1,0.4,1.4,0 c0.4-0.4,0.4-1,0-1.4l-2.8-2.8c-0.4-0.4-1-0.4-1.4,0c-0.4,0.4-0.4,1,0,1.4L12.5,14.1z M54.5,9.9l-2.8,2.8c-0.4,0.4-0.4,1,0,1.4 c0.4,0.4,1,0.4,1.4,0l2.8-2.8c0.4-0.4,0.4-1,0-1.4C55.6,9.5,54.9,9.5,54.5,9.9z M12.5,51.9l-2.8,2.8c-0.4,0.4-0.4,1,0,1.4 c0.4,0.4,1,0.4,1.4,0l2.8-2.8c0.4-0.4,0.4-1,0-1.4C13.6,51.5,12.9,51.5,12.5,51.9z M53.1,51.9c-0.4-0.4-1-0.4-1.4,0 c-0.4,0.4-0.4,1,0,1.4l2.8,2.8c0.4,0.4,1,0.4,1.4,0c0.4-0.4,0.4-1,0-1.4L53.1,51.9z M33,60c-0.6,0-1,0.4-1,1v4c0,0.6,0.4,1,1,1 s1-0.4,1-1v-4C34,60.4,33.6,60,33,60z M65,32h-4c-0.6,0-1,0.4-1,1s0.4,1,1,1h4c0.6,0,1-0.4,1-1S65.6,32,65,32z M5,32H1 c-0.6,0-1,0.4-1,1s0.4,1,1,1h4c0.6,0,1-0.4,1-1S5.6,32,5,32z"
              fill="#ffffff"
            />
          </svg>
          <div
            class="progress"
          >
            <span
              class="count"
            >
              30
            </span>
          </div>
        </div>
        <p
          class="genEnergy"
        >
          <svg
            class="genEnergyIcon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 21h-1l1-7H7.5c-.58 0-.57-.32-.38-.66.19-.34.05-.08.07-.12C8.48 10.94 10.42 7.54 13 3h1l-1 7h3.5c.49 0 .56.33.47.51l-.07.15C12.96 17.55 11 21 11 21z"
            />
          </svg>
           0
        </p>
      </div>
      <div
        class="ready"
      >
        <p>
          制限時間内に画面を連打してバッテリーを充電してください。
        </p>
        <button
          class="button --basic"
          type="button"
        >
          準備OK はじめる
        </button>
      </div>
    </div>
    <footer
      class="footer"
    >
      <button
        class="button --basic"
        disabled=""
        type="button"
      >
        救助ミッションへ
      </button>
    </footer>
  </article>
</DocumentFragment>
`;
